### PR TITLE
Add explanation to modelling exercise

### PR DESCRIPTION
**Pull request by iozbeyli**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin)
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Description
Added explanation for modeling exercise creation, modeling exercise example submission and modeling exercise submission

### Steps for Testing
Note: Example Submission is not being listed is not related to this pull request.

1. Create a modeling exercise give example solution explanation, create example submission with explanation
2. Log in as student and create a submission with explanation
3. Log in as tutor and check if you can see if the user can see example solution explanation and explanation for submissions in assessment
4. Assessing the explanation will be added with followup pull request
5. Follow same steps for modeling exercise in an exam
6. Check if explanation is missing anywhere modeling exercise submission related 

### Test Coverage
modeling-exam-submission.component.ts 96.97%
modeling-exam-summary.component.ts 100%
modeling-explanation-editor.component.ts 100%
modeling-assessment.component.ts  78.63%

#### Modeling Exercise Creation with Example Solution Explanation
out-19
#### Example Submission with Explanation for Modeling Exercise
out-20
#### Modeling Exercise Submission with Explanation
out-21
#### Modeling Exercise Assessment: Example Solution with Explanation
out-22_1
#### Modeling Exercise Assessment: Submission with Explanation
out-23
#### Modeling Exercise in Exam with Explanation
out-24
#### Team Exercise with Explanation
out-26

### CODE REVIEW

**Review by tiffca — APPROVED** [review]
Nice changes, worked well overall. I noticed 2 points, but maybe not critical, hence approving 

- Maybe this was only an error for me and someone else can please try to check it as well, however I couldn't view the "Example submission" while assessing. It only showed me this screen. 
image
Reproduce steps: 
1. Create a modeling exercise
2. Create an example submission
3. Submit solutions as a student 
4. Go to the exercise dashboard and try to view the example submission 
5. Actual: can't view, says there are 0 example submissions. Expected: should be able to view the example submission 
 

- For the exam, the explanation isn't viewable in the student's view at the end of the exam. So it might appear to the student that it wasn't saved. 

image

hint for other testers: make sure to refresh the browser if you can't view the explanation window immediately

**Review by airelawaleria — COMMENTED** [review]
Tested in the testing session with @simonzilker @marvinchimeka @JohannesWeiss @EneaGore . The issue mentioned by @tiffca

**Review by simonzilker — COMMENTED** [review]
Tested in testing session. Noticed the same issue for exams what @tiffca said. In addition, from the student account I could not see the before added explanation after my submission was graded.

**Review by JohannesWeiss — COMMENTED** [review]
Tested on ts1 in testing session. Explanations do not show up in exam's overview and review. Tutors can see them during the assessment, though. Basically the same issue as mentioned above.

**Review by EneaGore — COMMENTED** [review]
Overall worked fine but same issues as already stated. Also there seems to be no limit to charachters on the Explanation.

**Review by marvinchimeka — COMMENTED** [review]
tested in testing session. Experienced the same issue as @tiffca.

**Review by airelawaleria — CHANGES_REQUESTED** [review]
Tested in the testing session with @simonzilker @marvinchimeka @JohannesWeiss @EneaGore . The issue mentioned by @tiffca for exams and also a student cannot view the explanations in the results view for a modelling exercise

**Review by valeriebucher — COMMENTED** [review]
Can confirm @tiffca issues. But I noticed that creating an „example solution“ instead of „example submission“ it then shows up like in the screenshots. So one can view the example with the explanation when starting the assessment as a tutor.

**Comment by iozbeyli** [comment]
Thank you all for your feedback 🙂

**Comment by tiffca** [comment]
It looks like explanation is also missing when you view your modeling exercise submission after the due date. 

image
1. After submission due date, click on view submission 
2. Notice the explanation is missing 

Not sure if this is intended or to be resolved in a follow-up PR

**Comment by FrankeLukas** [comment]
> Also, during assessing it still shows me 0 example submissions, maybe someone else can also check.

I can't see any example submissions either. Not sure what is going on.

File: src/test/javascript/spec/component/exam/participate/summary/modeling-exam-summary.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            });
+    });
+
+    it('should initialize', () => {
+        fixture.detectChanges();
+        expect(ModelingExamSummaryComponent).to.be.ok;
+    });
+
+    it('should show no submission when there is no uml model', () => {
+        comp.ngOnInit();
```
Comment (stefanwaldhauser) [inline src/test/javascript/spec/component/exam/participate/summary/modeling-exam-summary.component.spec.ts:38]: Not a big deal but this is not necessary as detectChanges will call ngOnInit.

File: src/test/javascript/spec/component/modeling-explanation-editor/modeling-explanation-editor.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    it('should initialize', () => {
+        fixture.detectChanges();
+        expect(ModelingExplanationEditorComponent).to.be.ok;
+    });
+
+    it('should change explanation value bidirectionally between component and template', () => {
+        comp.explanation = 'Initial Explanation';
+        fixture.detectChanges();
+        fixture.whenStable().then(() => {
```
Comment (FrankeLukas) [inline src/test/javascript/spec/component/modeling-explanation-editor/modeling-explanation-editor.component.spec.ts:41]: Why is this necessary? I thought `detectChanges` already takes care of "stabilising" the component.

**Review by FrankeLukas — DISMISSED** [review]
Lgtm otherwise.

**Review by stefanwaldhauser — COMMENTED** [review]
Works well except the issues mentioned above

File: src/test/javascript/spec/component/modeling-explanation-editor/modeling-explanation-editor.component.spec.ts (hunk 2 above)
Comment (iozbeyli) [reply src/test/javascript/spec/component/modeling-explanation-editor/modeling-explanation-editor.component.spec.ts:41]: Without it textarea value does not get updated, I believe it is related to FormsModule.

Comment (FrankeLukas) [reply src/test/javascript/spec/component/modeling-explanation-editor/modeling-explanation-editor.component.spec.ts:41]: Ah ok.

File: src/test/javascript/spec/component/exam/participate/summary/modeling-exam-summary.component.spec.ts (hunk 1 above)
Comment (iozbeyli) [reply src/test/javascript/spec/component/exam/participate/summary/modeling-exam-summary.component.spec.ts]: Yeap, removed those.

**Comment by iozbeyli** [comment]
> It looks like explanation is also missing when you view your modeling exercise submission after the due date.
> 
> image
> 
>     1. as a student, After submission due date, click on view submission
> 
>     2. Notice the explanation is missing
> 
> 
> Not sure if this is intended or to be resolved in a follow-up PR.
> 
> Also, during assessing it still shows me 0 example submissions, maybe someone else can also check.

It should be fixed now.

**Review by philgabr — APPROVED** [review]
I just tested it (for the first time) on ts1 and everything seems to work as intended :+1:

**Review by EneaGore — APPROVED** [review]
Tested in testing session, approve.

**Review by simonzilker — APPROVED** [review]
Tested in testing session. Works perfectly!

**Review by JohannesWeiss — APPROVED** [review]
Tested in testing session on ts1, worked as expected. The previous issue is resolved now.

**Review by airelawaleria — APPROVED** [review]
Tested in the testing session with @simonzilker @marvinchimeka @JohannesWeiss @EneaGore @LudwigWeide . Works as expected

**Review by marvinchimeka — APPROVED** [review]
Tested in testing session, worked as intended.

**Review by tiffca — APPROVED** [review]
The before mentioned issues look resolved, nice changes!

**Review by krusche — APPROVED** [review]

**Comment by krusche** [comment]
Thanks for the nice contribution!
